glide-core/redis-rs/redis/src/cluster_async/pipeline_routing.rs

```diff
@@ -0,0 +1,433 @@
+///
+/// This function updates the `values_and_addresses` vector at the given `index` and optionally at the
+/// `inner_index` if provided. If `inner_index` is `Some`, it ensures that the vector at that index is large enough
+/// to hold the value and address at the specified position, resizing it if necessary. If `inner_index` is `None`,
+/// the value and address are simply appended to the vector.
+///
+/// # Parameters
+/// - `values_and_addresses`: A mutable reference to a vector of vectors that stores the results of pipeline commands.
+/// - `index`: The index in `values_and_addresses` where the result should be stored.
+/// - `inner_index`: An optional index within the vector at `index`, used to store the result at a specific position.
+/// - `value`: The result value to store.
+/// - `address`: The address associated with the result.
+pub fn add_pipeline_result(
+    values_and_addresses: &mut [Vec<(Value, String)>],
+    index: usize,
+    inner_index: Option<usize>,
+    value: Value,
+    address: String,
+) {
+    match inner_index {
+        Some(inner_index) => {
+            // Ensure the vector at the given index is large enough to hold the value and address at the specified position
+            if values_and_addresses[index].len() <= inner_index {
+                values_and_addresses[index].resize(inner_index + 1, (Value::Nil, "".to_string()));
+            }
+            values_and_addresses[index][inner_index] = (value, address);
+        }
+        None => values_and_addresses[index].push((value, address)),
+    }
+}
+
+/// Collects and processes the results of pipeline tasks from a `tokio::task::JoinSet`.
+///
+/// This function iteratively retrieves completed tasks from the provided `join_set` and processes
+/// their results. Successful results are added to the `values_and_addresses` vector using the
+/// indices and values provided. If an error occurs in any task, it is recorded and returned as
+/// the first encountered error.
+///
+/// # Parameters
+/// - `join_set`: A mutable reference to a `tokio::task::JoinSet` containing tasks that return:
+///   - `Ok((Vec<(usize, Option<usize>)>, Vec<Value>, String))`: On success, a tuple of:
+///     - A list of indices and optional inner indices corresponding to pipeline commands.
+///     - A list of `Value` results from the executed pipeline.
+///     - The `String` address where the task was executed.
+///   - `Err((OperationTarget, RedisError))`: On failure, an error detailing the operation target and the Redis error.
+/// - `values_and_addresses`: A mutable slice of vectors, where each vector corresponds to a pipeline
+///   command's results. This is updated with the values and addresses from successful tasks.
+///
+/// # Returns
+/// - `Ok(Some((OperationTarget, RedisError)))`: If one or more tasks encountered an error, returns the first error.
+/// - `Ok(None)`: If all tasks completed successfully.
```

Comment:
since returning Ok with Some(err) is confusing, you can make it more readable with some enum representing the return values, something like 
```
pub enum MultiPipelineResult {
    /// All tasks completed successfully.
    AllSuccessful,

    /// Some tasks failed, returning the first encountered error and the associated operation target.
    FirstError {
        target: OperationTarget,
        error: RedisError,
    },
}